File: schema/events.ts
/**
 * 统一测试事件模型 — TypeScript 类型定义
 *
 * 所有测试框架（Vitest / gdUnit4 / Playwright / Airtest）
 * 产生的事件都转换为这个格式。
 */

// ─── 事件类型枚举 ──────────────────────────────────────────

/** 测试生命周期事件 */
export type TestLifecycleType =
  | "test.start"
  | "test.end"
  | "test.skip"
  | "test.fail"
  | "test.error";

/** 断言事件 */
export type AssertionType =
  | "assert.pass"
  | "assert.fail"
  | "assert.semantic";

/** 用户操作事件 */
export type ActionType =
  | "action.click"
  | "action.input"
  | "action.navigate"
  | "action.wait"
  | "action.screenshot";

/** 游戏特有事件 */
export type GameEventType =
  | "game.state_change"
  | "game.scene_load"
  | "game.signal_emit"
  | "game.save"
  | "game.load";

/** Agent 特有事件 */
export type AgentEventType =
  | "agent.llm_call"
  | "agent.tool_call"
  | "agent.tool_result"
  | "agent.thinking";

/** 观测事件 */
export type ObservationType =
  | "observation.snapshot"
  | "observation.coverage"
  | "observation.anomaly";

/** 报告事件 */
export type ReportType =
  | "report.summary"
  | "report.bug_candidate"
  | "report.gate_result";

/** 所有事件类型 */
export type ObsEventType =
  | TestLifecycleType
  | AssertionType
  | ActionType
  | GameEventType
  | AgentEventType
  | ObservationType
  | ReportType;

// ─── 框架和项目标识 ────────────────────────────────────────

export type Framework =
  | "vitest"
  | "gdunit4"
  | "playwright"
  | "airtest"
  | "custom";

export type Project =
  | "travel-agent"
  | "pogongshichongzou"
  | "loopexpedition"
  | string; // 允许扩展

// ─── 事件来源 ──────────────────────────────────────────────

export interface EventSource {
  framework: Framework;
  project: Project;
  file?: string;
  test_name?: string;
  suite?: string;
}

// ─── 核心事件结构 ──────────────────────────────────────────

export interface ObsEvent {
  /** 唯一标识 */
  event_id: string;

  /** 测试会话 ID */
  session_id: string;

  /** Unix 毫秒时间戳 */
  timestamp: number;

  /** 事件来源 */
  source: EventSource;

  /** 事件类型 */
  type: ObsEventType;

  /** 事件数据（不同类型有不同的结构） */
  data: Record<string, unknown>;

  /** 父事件 ID（用于嵌套/因果链） */
  parent_event_id?: string;

  /** 跨系统追踪 ID */
  trace_id?: string;

  /** 追踪 span ID */
  span_id?: string;
}

// ─── 事件数据类型（按事件类型分） ─────────────────────────

/** test.start 数据 */
export interface TestStartData {
  full_name: string;
  tags?: string[];
}

/** test.end / test.fail 数据 */
export interface TestEndData {
  duration_ms: number;
  state: "pass" | "fail" | "skip";
  errors?: Array<{
    message: string;
    stack?: string;
    expected?: unknown;
    actual?: unknown;
  }>;
  assertions?: Array<{
    name: string;
    state: "pass" | "fail";
    duration?: number;
  }>;
}

/** assert.* 数据 */
export interface AssertionData {
  assertion_name: string;
  passed: boolean;
  expected?: unknown;
  actual?: unknown;
  message?: string;
  details?: Record<string, unknown>;
}

/** assert.semantic 数据（AI 语义断言） */
export interface SemanticAssertionData extends AssertionData {
  evaluator: string;
  score: number; // 0-1
  dimensions?: Record<string, number>;
  llm_model?: string;
  llm_reasoning?: string;
}

/** action.* 数据 */
export interface ActionData {
  target?: string;
  value?: unknown;
  duration_ms?: number;
  screenshot_base64?: string;
  metadata?: Record<string, unknown>;
}

/** game.state_change 数据 */
export interface GameStateData {
  scene_path?: string;
  node_path?: string;
  state: Record<string, unknown>;
  previous_state?: Record<string, unknown>;
}

/** game.signal_emit 数据 */
export interface GameSignalData {
  signal_name: string;
  source_node: string;
  args?: unknown[];
}

/** agent.llm_call 数据 */
export interface LlmCallData {
  model: string;
  input_tokens?: number;
  output_tokens?: number;
  duration_ms?: number;
  prompt?: string;
  response?: string;
  error?: string;
}

/** agent.tool_call 数据 */
export interface ToolCallData {
  tool_name: string;
  input: unknown;
  cost_tier?: string;
}

/** agent.tool_result 数据 */
export interface ToolResultData {
  tool_name: string;
  input: unknown;
  output: unknown;
  duration_ms?: number;
  success: boolean;
  error?: string;
}

/** observation.snapshot 数据 */
export interface SnapshotData {
  tree_summary?: string;
  runtime_state?: Record<string, unknown>;
  has_screenshot: boolean;
  screenshot_base64?: string;
}

/** observation.coverage 数据 */
export interface CoverageData {
  seen_events: string[];
  seen_obs_keys: string[];
  event_coverage: number; // 0-1
  obs_coverage: number; // 0-1
  total_flows?: number;
  covered_flows?: number;
  uncovered_flows?: string[];
}

/** observation.anomaly 数据 */
export interface AnomalyData {
  severity: "high" | "medium" | "low";
  category: string;
  description: string;
  step?: number;
  evidence: Record<string, unknown>;
}

/** report.summary 数据 */
export interface SummaryData {
  total_tests: number;
  passed: number;
  failed: number;
  skipped: number;
  duration_ms: number;
  pass_rate: number;
}

/** report.bug_candidate 数据 */
export interface BugCandidateData {
  severity: "high" | "medium" | "low";
  category: string;
  description: string;
  step?: number;
  evidence: Record<string, unknown>;
  confidence: number;
}

/** report.gate_result 数据 */
export interface GateResultData {
  verdict: "PASS" | "FAIL";
  rules: Record<string, {
    value: number;
    threshold: number;
    comparator: string;
    pass: boolean;
  }>;
}

// ─── 测试会话 ──────────────────────────────────────────────

export interface ObsSession {
  session_id: string;
  project: Project;
  framework: Framework;
  started_at: number;
  ended_at?: number;
  total_tests?: number;
  passed_tests?: number;
  failed_tests?: number;
  duration_ms?: number;
  gate_result?: GateResultData;
  metadata?: Record<string, unknown>;
}

// ─── AI 分析结果 ───────────────────────────────────────────

export interface AnalysisResult {
  analysis_id: string;
  session_id: string;
  timestamp: number;
  analyzer: string;
  findings: Array<Record<string, unknown>>;
  confidence: number;
  summary: string;
  recommendations: string[];
}

// ─── Timeline 可视化格式 ───────────────────────────────────

export interface TimelineItem {
  id: string;
  group: string;
  start: number;
  end?: number;
  content: string;
  className?: string;
  data: ObsEvent;
}

export interface TimelineGroup {
  id: string;
  content: string;
  subgroupOrder?: string;
}

File: schema/events.py
"""统一测试事件模型 — Python 类型定义

所有测试框架（Vitest / gdUnit4 / Playwright / Airtest）
产生的事件都转换为这个格式。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional

# ─── 事件类型枚举 ──────────────────────────────────────────


class TestLifecycleType(str, Enum):
    TEST_START = "test.start"
    TEST_END = "test.end"
    TEST_SKIP = "test.skip"
    TEST_FAIL = "test.fail"
    TEST_ERROR = "test.error"


class AssertionType(str, Enum):
    ASSERT_PASS = "assert.pass"
    ASSERT_FAIL = "assert.fail"
    ASSERT_SEMANTIC = "assert.semantic"


class ActionType(str, Enum):
    ACTION_CLICK = "action.click"
    ACTION_INPUT = "action.input"
    ACTION_NAVIGATE = "action.navigate"
    ACTION_WAIT = "action.wait"
    ACTION_SCREENSHOT = "action.screenshot"


class GameEventType(str, Enum):
    GAME_STATE_CHANGE = "game.state_change"
    GAME_SCENE_LOAD = "game.scene_load"
    GAME_SIGNAL_EMIT = "game.signal_emit"
    GAME_SAVE = "game.save"
    GAME_LOAD = "game.load"


class AgentEventType(str, Enum):
    AGENT_LLM_CALL = "agent.llm_call"
    AGENT_TOOL_CALL = "agent.tool_call"
    AGENT_TOOL_RESULT = "agent.tool_result"
    AGENT_THINKING = "agent.thinking"


class ObservationType(str, Enum):
    OBSERVATION_SNAPSHOT = "observation.snapshot"
    OBSERVATION_COVERAGE = "observation.coverage"
    OBSERVATION_ANOMALY = "observation.anomaly"
    OBSERVATION_STATE_DIFF = "observation.state_diff"
    OBSERVATION_INPUT_TRACE = "observation.input_trace"
    OBSERVATION_CAUSE_TRACE = "observation.cause_trace"


class DebugEventType(str, Enum):
    """调试协议事件 (借鉴 OpenGame Debug Skill)"""

    DEBUG_ITERATION = "debug.iteration"
    DEBUG_MATCH = "debug.match"
    DEBUG_REPAIR = "debug.repair"
    DEBUG_EVOLVE = "debug.evolve"


class BenchEventType(str, Enum):
    """评估事件 (借鉴 OpenGame-Bench)"""

    BENCH_BUILD_HEALTH = "bench.build_health"
    BENCH_VISUAL_USABILITY = "bench.visual_usability"
    BENCH_INTENT_ALIGNMENT = "bench.intent_alignment"
    BENCH_RESULT = "bench.result"


class ReportType(str, Enum):
    REPORT_SUMMARY = "report.summary"
    REPORT_BUG_CANDIDATE = "report.bug_candidate"
    REPORT_GATE_RESULT = "report.gate_result"


# 所有事件类型的联合（用于类型提示）
ObsEventType = str  # 实际值为上述枚举的 value


# ─── 框架和项目标识 ────────────────────────────────────────


class Framework(str, Enum):
    VITEST = "vitest"
    GDUNIT4 = "gdunit4"
    PLAYWRIGHT = "playwright"
    AIRTEST = "airtest"
    GODOT_E2E = "godot_e2e"
    GODOT_DRIVER = "godot_driver"
    CUSTOM = "custom"


# ─── 核心数据结构 ──────────────────────────────────────────


@dataclass
class EventSource:
    """事件来源"""

    framework: str  # Framework 枚举值
    project: str
    file: Optional[str] = None
    test_name: Optional[str] = None
    suite: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        d = {
            "framework": self.framework,
            "project": self.project,
        }
        if self.file:
            d["file"] = self.file
        if self.test_name:
            d["test_name"] = self.test_name
        if self.suite:
            d["suite"] = self.suite
        return d


@dataclass
class ObsEvent:
    """统一测试事件"""

    event_id: str
    session_id: str
    timestamp: int  # Unix ms
    source: EventSource
    type: str  # ObsEventType 枚举值
    data: Dict[str, Any]
    parent_event_id: Optional[str] = None
    trace_id: Optional[str] = None
    span_id: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        d: Dict[str, Any] = {
            "event_id": self.event_id,
            "session_id": self.session_id,
            "timestamp": self.timestamp,
            "source": self.source.to_dict(),
            "type": self.type,
            "data": self.data,
        }
        if self.parent_event_id:
            d["parent_event_id"] = self.parent_event_id
        if self.trace_id:
            d["trace_id"] = self.trace_id
        if self.span_id:
            d["span_id"] = self.span_id
        return d


@dataclass
class ObsSession:
    """测试会话"""

    session_id: str
    project: str
    framework: str
    started_at: int  # Unix ms
    ended_at: Optional[int] = None
    total_tests: Optional[int] = None
    passed_tests: Optional[int] = None
    failed_tests: Optional[int] = None
    duration_ms: Optional[int] = None
    gate_result: Optional[Dict[str, Any]] = None
    metadata: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        d: Dict[str, Any] = {
            "session_id": self.session_id,
            "project": self.project,
            "framework": self.framework,
            "started_at": self.started_at,
        }
        if self.ended_at:
            d["ended_at"] = self.ended_at
        if self.total_tests is not None:
            d["total_tests"] = self.total_tests
        if self.passed_tests is not None:
            d["passed_tests"] = self.passed_tests
        if self.failed_tests is not None:
            d["failed_tests"] = self.failed_tests
        if self.duration_ms is not None:
            d["duration_ms"] = self.duration_ms
        if self.gate_result:
            d["gate_result"] = self.gate_result
        if self.metadata:
            d["metadata"] = self.metadata
        return d


@dataclass
class AnalysisResult:
    """AI 分析结果"""

    analysis_id: str
    session_id: str
    timestamp: int
    analyzer: str
    findings: List[Dict[str, Any]]
    confidence: float  # 0-1
    summary: str
    recommendations: List[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "analysis_id": self.analysis_id,
            "session_id": self.session_id,
            "timestamp": self.timestamp,
            "analyzer": self.analyzer,
            "findings": self.findings,
            "confidence": self.confidence,
            "summary": self.summary,
            "recommendations": self.recommendations,
        }


# ─── 事件工厂函数 ──────────────────────────────────────────

import time
import uuid


def _generate_id() -> str:
    return str(uuid.uuid4())


def _timestamp() -> int:
    return int(time.time() * 1000)


def create_test_start(
    session_id: str,
    source: EventSource,
    test_name: str,
    full_name: str,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建 test.start 事件"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=TestLifecycleType.TEST_START.value,
        data={"test_name": test_name, "full_name": full_name},
        trace_id=trace_id,
    )


def create_test_end(
    session_id: str,
    source: EventSource,
    test_name: str,
    passed: bool,
    duration_ms: int,
    errors: Optional[List[Dict[str, Any]]] = None,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建 test.end / test.fail 事件"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=TestLifecycleType.TEST_END.value if passed else TestLifecycleType.TEST_FAIL.value,
        data={
            "test_name": test_name,
            "passed": passed,
            "duration_ms": duration_ms,
            "errors": errors or [],
        },
        trace_id=trace_id,
    )


def create_assertion(
    session_id: str,
    source: EventSource,
    assertion_name: str,
    passed: bool,
    expected: Any = None,
    actual: Any = None,
    message: Optional[str] = None,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建 assert.* 事件"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=AssertionType.ASSERT_PASS.value if passed else AssertionType.ASSERT_FAIL.value,
        data={
            "assertion_name": assertion_name,
            "passed": passed,
            "expected": expected,
            "actual": actual,
            "message": message,
        },
        trace_id=trace_id,
    )


def create_agent_tool_call(
    session_id: str,
    source: EventSource,
    tool_name: str,
    tool_input: Any,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建 agent.tool_call 事件"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=AgentEventType.AGENT_TOOL_CALL.value,
        data={
            "tool_name": tool_name,
            "input": tool_input,
        },
        trace_id=trace_id,
    )


def create_agent_tool_result(
    session_id: str,
    source: EventSource,
    tool_name: str,
    tool_input: Any,
    tool_output: Any,
    duration_ms: int,
    success: bool,
    error: Optional[str] = None,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建 agent.tool_result 事件"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=AgentEventType.AGENT_TOOL_RESULT.value,
        data={
            "tool_name": tool_name,
            "input": tool_input,
            "output": tool_output,
            "duration_ms": duration_ms,
            "success": success,
            "error": error,
        },
        trace_id=trace_id,
    )


def create_game_state_change(
    session_id: str,
    source: EventSource,
    scene_path: str,
    state: Dict[str, Any],
    previous_state: Optional[Dict[str, Any]] = None,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建 game.state_change 事件"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=GameEventType.GAME_STATE_CHANGE.value,
        data={
            "scene_path": scene_path,
            "state": state,
            "previous_state": previous_state,
        },
        trace_id=trace_id,
    )


def create_bug_candidate(
    session_id: str,
    source: EventSource,
    severity: str,
    category: str,
    description: str,
    evidence: Dict[str, Any],
    confidence: float = 0.8,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建 report.bug_candidate 事件"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=ReportType.REPORT_BUG_CANDIDATE.value,
        data={
            "severity": severity,
            "category": category,
            "description": description,
            "evidence": evidence,
            "confidence": confidence,
        },
        trace_id=trace_id,
    )


def create_gate_result(
    session_id: str,
    source: EventSource,
    verdict: str,
    rules: Dict[str, Any],
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建 report.gate_result 事件"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=ReportType.REPORT_GATE_RESULT.value,
        data={
            "verdict": verdict,
            "rules": rules,
        },
        trace_id=trace_id,
    )


# ─── Godot 游戏测试事件工厂 ──────────────────────────────────


def create_visual_assertion(
    session_id: str,
    source: EventSource,
    template_name: str,
    matched: bool,
    confidence: float = 0.0,
    position: Optional[List[int]] = None,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建视觉断言事件 (Airtest 风格)"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=AssertionType.ASSERT_PASS.value if matched else AssertionType.ASSERT_FAIL.value,
        data={
            "assertion_type": "visual_template",
            "template_name": template_name,
            "matched": matched,
            "confidence": confidence,
            "position": position,
        },
        trace_id=trace_id,
    )


def create_debug_event(
    session_id: str,
    source: EventSource,
    debug_type: str,
    entry_id: Optional[str] = None,
    error_code: Optional[str] = None,
    error_message: Optional[str] = None,
    fix_description: Optional[str] = None,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建调试协议事件 (OpenGame Debug Skill 风格)"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=debug_type,
        data={
            "entry_id": entry_id,
            "error_code": error_code,
            "error_message": error_message,
            "fix_description": fix_description,
        },
        trace_id=trace_id,
    )


def create_bench_event(
    session_id: str,
    source: EventSource,
    dimension: str,
    score: float,
    passed: bool,
    checks: Optional[List[Dict[str, Any]]] = None,
    trace_id: Optional[str] = None,
) -> ObsEvent:
    """创建评估事件 (OpenGame-Bench 风格)"""
    return ObsEvent(
        event_id=_generate_id(),
        session_id=session_id,
        timestamp=_timestamp(),
        source=source,
        type=f"bench.{dimension}",
        data={
            "dimension": dimension,
            "score": score,
            "passed": passed,
            "checks": checks or [],
        },
        trace_id=trace_id,
    )
